### PR TITLE
Support atom and string field names in form inputs

### PR DESCRIPTION
**Pull request by miguel-s**
Related to #321 and in response to https://elixirforum.com/t/dynamically-generated-forms-and-running-out-of-input-atoms/38582.

Phoenix HTML form helpers allow atoms and strings for field names, so we should, too.

### CODE REVIEW

File: lib/surface/components/form/error_tag.ex [hunk 1]
```diff
@@ -106,7 +106,7 @@ defmodule Surface.Components.Form.ErrorTag do
     ~H"""
     <InputContext assigns={{ assigns }} :let={{ form: form, field: field }}>
       <span
-        :for={{ error <- Keyword.get_values(form.errors, field) }}
+        :for={{ error <- Keyword.get_values(form.errors, String.to_existing_atom(field)) }}
```
Comment (miguel-s) [inline lib/surface/components/form/error_tag.ex:109]: Is it safe to cast strings as atoms here?

File: test/components/form_test.exs [hunk 2]
```diff
@@ -20,7 +20,7 @@ defmodule Surface.Components.FormTest do
     def render(assigns) do
       ~H"""
       <Form for={{ @changeset }} action="#" csrf_token="test" as={{ :user }} :let={{ form: f }}>
-        <TextInput field="name" />
+        <TextInput field={{ :name }} />
```
[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: Yes, form with changeset seems to expect atom field names https://github.com/phoenixframework/phoenix_ecto/blob/54623ef9ead9053b50f8079618f1973e73dfc2f4/lib/phoenix_ecto/html.ex#L105

[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: Getting the same error when using form_for with a changeset and using string field names.
Controller
```Elixir
def index(conn, _params) do
  changeset = Accounts.change_category(%User{})
  render(conn, "index.html", changeset: changeset)
end
```
Template
```Eex
<%= form_for @changeset, "#", fn f -> %>
  <label>
    Name: <%= text_input f, "name" %>
  </label>
  <%= submit "Submit" %>
<% end %>
```
Error
> ** (ArgumentError) argument error
        :erlang.atom_to_binary("name", :utf8)
        (phoenix_ecto 4.2.1) lib/phoenix_ecto/html.ex:111: Phoenix.HTML.FormData.Ecto.Changeset.input_value/4
        (phoenix_html 2.14.2) lib/phoenix_html/form.ex:482: Phoenix.HTML.Form.input_value/2
        (phoenix_html 2.14.2) lib/phoenix_html/form.ex:825: Phoenix.HTML.Form.generic_input/4

[the remark this answers is not in the document]
Comment (msaraiva) [reply test/components/form_test.exs:23]: That might be the reason we ended up using atoms in the first place 🤔

[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: It doesn't look like this is a big issue in Phoenix (haven't found any issues in `phoenix_html` or `phoenix_ecto` about this).

However, I see one risk in Surface:
1.  Phoenix HTLM encourages the use atoms:
  - Documentation shows examples with atoms
  - Form helpers look"natural" with atoms `<%= text_input f, :name %>`
2. Surface encourages the use of strings:
  - Documentation shows examples with strings
  - It looks more "natural" to write `<Field name="name">` than `<Field name={{ :name }}>`

So even if we offer the same behaviour as Phoenix HTML, I think Surface users will trip over this error more often.
Do you agree this could be an issue? If so, what can we do to mitigate this?

[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: Once this https://github.com/phoenixframework/phoenix_ecto/pull/139 is released we'll get a better error message:
`** (ArgumentError) expected field to be an atom, got: "name"`

I believe that should be enough.

[the remark this answers is not in the document]
Comment (msaraiva) [reply test/components/form_test.exs:23]: Without the changes in this PR, whenever the user passes a literal string, like:

```
<Field name="name">
```

it will automatically convert it to an `atom` since the type is `:atom`. We do this here: https://github.com/surface-ui/surface/blob/ed77d6b359ef29df9dabf78b147301ac04dad17c/lib/surface/type_handler/atom.ex#L7-L9

It's nice but probably not a good idea as I thought it was when I implemented it.

By changing the type to `:any`, this conversion will no longer happen. I believe this may introduce a breaking change. Correct? 

Do you think removing this automatic conversion would help to avoid confusion? If we do that, the user will always have to write `name={{  :name }}` for atoms and `name="name"` for strings. 

Note: If we change this, we should probably change the type `:form` too as does the same string-to-atom conversion here: https://github.com/surface-ui/surface/blob/ed77d6b359ef29df9dabf78b147301ac04dad17c/lib/surface/type_handler/form.ex#L7-L9

Also, before removing this behaviour, maybe we could just emit a warning. Something like:

```
<Form name="myform">
```
```
Warning: support to set the form's name as string is deprecated and will be removed soon. 
Please, update your code to use `name={{ :myform }}` instead of `name="myform"`.
```

Do it make sense?

[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: Yes changing the type to `:any` is technically a breaking change, but based on our tests it seems to really only break forms with changeset and string keys, other cases passed the tests.

I personally like the auto conversion, it allows for more html-like templates while still using atoms (which I prefer in general). It's true that it is a bit of magic, but declaring the type of the prop as `:atom` was enough of a hint for me.

Regarding form helpers with string name, phoenix_ecto accepted my PR phoenixframework/phoenix_ecto#139 to improve the error message (looks like this now `** (ArgumentError) expected field to be an atom, got: "name")`. I believe this should be enough to handle this edge case?

[the remark this answers is not in the document]
Comment (msaraiva) [reply test/components/form_test.exs:23]: > I personally like the auto conversion, it allows for more html-like templates while still using atoms (which I prefer in general). It's true that it is a bit of magic, but declaring the type of the prop as `:atom` was enough of a hint for me.

The problem I see with auto-converting string-to-atom it's because it's nice but not intuitive. Possible reasons:

1. By looking at it in the template, it gives the impression the value is string, not atom. That means for every attribute passed as `atrr="string"`, the user needs to look into its type definition to figure out the type, which will be most likely in a separate file. Tools like mouse-over hints provided by editors might help but we shouldn't assume the user has that.
2. It only works for literals, which makes this even more confusing. We know the reason is security, the user might not know that, especially newcomers. Imagine a code where you are passing a literal string to an attribute of type `:atom`, then later on you change the implementation to pass a variable or assign. Intuitively, you think the value is a string (after all you were passing a string before) so you implement the changes assuming that variable/assign must be a string. Now you have a bug which you'll probably only notice at runtime.  

Maybe it would be nice to get other opinions about this, especially because it's a breaking change.

@surface-ui/core what are your thoughts about dropping auto-conversion of attributes values passed as string literals to atom?

[the remark this answers is not in the document]
Comment (miguel-s) [reply test/components/form_test.exs:23]: Yes those are good arguments, we should probably remove it.

Would it be possible (or desirable) to add a shorthand form for atoms like `<Form name=:myform>`?

**Review by msaraiva — APPROVED** [review]
@miguel-s I think we can merge this as it is. Please don't forget to add this change to the CHANGELOG as we'll have to make it clear it's a breaking change. We can move the discussion about the auto-conversion stuff to someplace else. 

Thanks for fixing this! ❤️

**Comment by miguel-s** [comment]
@msaraiva Updated the CHANGELOG, if the line looks good to you I'll merge

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,5 +1,9 @@
 # Changelog
 
+## v0.4.0 (WIP)
+  
+  * Added support for string keys in form fields and removed autoconversion of strings to atoms (#319)
```
Comment (msaraiva) [inline CHANGELOG.md:5]: We actually didn't remove the autoconversion in general, right? It was more like a side-effect of changing the type to `:any`. So I believe something like this might be more useful to the user:

```
Support defining form fields as strings. Consequently, fields defined as literal strings will no 
longer be auto-converted to `:atom` and will keep the original value (#319)
```

**Note:** I think the changelog should be just the initial place to check for breaking changes, however, we should probably write a short migration guide giving some basic instructions on how to fix those breaking changes. That's usually not necessary but for `v0.4` might be a good idea as we're introducing a few of them.

@surface-ui/core WDYT?

Comment (miguel-s) [reply CHANGELOG.md:5]: Your version is definitely better!

**Review by msaraiva — APPROVED** [review]
